File: leetcode_problems/1_two_sum/two_sum_hash_map.java
class Solution {
    // @COMPANY Facebook
	// @KEYWORD Hash Map
    // @NOTE Using target-nums[i] as key.
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> hash = new HashMap<>();
        
        for (int i = 0; i < nums.length; i ++) {
            if (hash.containsKey(target - nums[i]))
                return new int[]{hash.get(target - nums[i]), i};
            hash.put(nums[i], i);
        }
        
        return new int[2];
    }
}

File: leetcode_problems/1_two_sum/two_sum_hash_map.cc
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> nums_table;
        vector<int> result;
        for (int i = 0; i < nums.size(); i ++){
            int complement = target - nums[i];
            if (nums_table.find(complement) != nums_table.end() && nums_table[complement] < i){
                result.push_back(nums_table[complement]);
                result.push_back(i);
                return(result);
            }
            nums_table[nums[i]] = i;
        }
    }
};